File: sandbox/src/bolt.ts
export const boltExample = `// NGK bolt: hex head, shank, core, and swept helical thread.
const { Point, Vector, Axis, Frame } = ngk.geometry;
const { polygon } = ngk.modeling.faces;
const { cylinder_at, extruded, fuse } = ngk.modeling.solids;
const { helix } = ngk.modeling.edges;
const { sweep_face } = ngk.modeling.sweep;

const TAU = 2 * Math.PI;
const HEAD_ACROSS_FLATS = 10;
const HEAD_HEIGHT = 4;
const BODY_RADIUS = 3;
const BODY_LENGTH = 8;
const CORE_RADIUS = 2.5;
const THREAD_LENGTH = 12;
const THREAD_PITCH = 1.25;
const THREAD_DEPTH = 0.5;
const THREAD_OVERLAP = 0.1;
const THREAD_BASE_WIDTH = 0.9;
const THREAD_MARGIN = 0.75;

function frameAt(z: number) {
  return new Frame(new Point(0, 0, z), new Vector(1, 0, 0), new Vector(0, 1, 0));
}

function head() {
  const radius = HEAD_ACROSS_FLATS / Math.sqrt(3);
  const corners = Array.from({ length: 6 }, (_, i) => {
    const angle = TAU * i / 6;
    return new Point(radius * Math.cos(angle), radius * Math.sin(angle), 0);
  });
  return extruded(polygon(corners), new Vector(0, 0, 1), HEAD_HEIGHT);
}

function body() {
  return cylinder_at(BODY_RADIUS, BODY_LENGTH, frameAt(-BODY_LENGTH));
}

function core() {
  return cylinder_at(CORE_RADIUS, THREAD_LENGTH, frameAt(-BODY_LENGTH - THREAD_LENGTH));
}

function thread() {
  const bottom = -BODY_LENGTH - THREAD_LENGTH + THREAD_MARGIN;
  const turns = (THREAD_LENGTH - 2 * THREAD_MARGIN) / THREAD_PITCH;
  const axis = new Axis(new Point(0, 0, bottom), new Vector(0, 0, 1));
  const spine = helix(axis, CORE_RADIUS, THREAD_PITCH, 0, TAU * turns);
  const start = spine.start.point;
  if (!start) throw new Error("Helix start has no point");
  const radialX = start.x / CORE_RADIUS;
  const radialY = start.y / CORE_RADIUS;
  const baseRadius = CORE_RADIUS - THREAD_OVERLAP;
  const crestRadius = CORE_RADIUS + THREAD_DEPTH;
  const half = THREAD_BASE_WIDTH / 2;
  const profile = polygon([
    new Point(radialX * baseRadius, radialY * baseRadius, start.z - half),
    new Point(radialX * crestRadius, radialY * crestRadius, start.z),
    new Point(radialX * baseRadius, radialY * baseRadius, start.z + half),
  ]);
  return sweep_face(profile, spine, axis, Math.ceil(turns * 16));
}

console.log("Building head and shank…");
let bolt = fuse(head(), body());
bolt = fuse(bolt, core());
console.log("Sweeping and fusing the thread…");
bolt = fuse(bolt, thread());
show(bolt);
console.log("Bolt ready: " + bolt.faceCount + " faces");
`;

File: examples/bolt.rs
//! Profiling workload: the hex bolt of `tests/parts/bolt.rs`, built and fused
//! without opening the viewer.
//!
//! A hexagonal head, a plain body, a thinner threaded core and a helical
//! thread are fused into one solid; the thread/core Boolean dominates.

use std::f64::consts::TAU;
use std::time::Instant;

use nalgebra::Vector3;
use ngk::builders::sweep::{SweepFrame, SweepOptions};
use ngk::geometry::{Axis3, Frame, Helix, NativeParam, Point3, SolverCounters};
use ngk::modeling::solids::{cylinder_at, extruded, fuse};
use ngk::modeling::sweep::sweep_face;
use ngk::modeling::{edges, faces};
use ngk::topology::shape::{Shape, SolidTag};
use ngk::topology::validation::{validate_solid_manifold, validate_solid_orientation};
use radians::Rad64;

const HEAD_ACROSS_FLATS: f64 = 10.0;
const HEAD_HEIGHT: f64 = 4.0;
const BODY_RADIUS: f64 = 3.0;
const BODY_LENGTH: f64 = 8.0;
const CORE_RADIUS: f64 = 2.5;
const THREAD_LENGTH: f64 = 12.0;
const THREAD_PITCH: f64 = 1.25;
const THREAD_DEPTH: f64 = 0.5;
const THREAD_OVERLAP: f64 = 0.1;
const THREAD_BASE_WIDTH: f64 = 0.9;
const THREAD_MARGIN: f64 = 0.75;
const SAMPLES_PER_TURN: usize = 16;

type Solid = Shape<SolidTag>;

fn head() -> Solid {
    let circumradius = HEAD_ACROSS_FLATS / 3.0_f64.sqrt();
    let corners: Vec<Point3> = (0..6)
        .map(|i| {
            let angle = TAU * i as f64 / 6.0;
            Point3::new(circumradius * angle.cos(), circumradius * angle.sin(), 0.0)
        })
        .collect();
    let hexagon = faces::polygon(&corners).expect("hexagon should build");
    extruded(hexagon, Vector3::z_axis(), HEAD_HEIGHT).expect("head should extrude")
}

fn body() -> Solid {
    cylinder_at(
        Frame::at(Point3::new(0.0, 0.0, -BODY_LENGTH)),
        BODY_RADIUS,
        BODY_LENGTH,
    )
    .expect("body should build")
}

fn core() -> Solid {
    cylinder_at(
        Frame::at(Point3::new(0.0, 0.0, -BODY_LENGTH - THREAD_LENGTH)),
        CORE_RADIUS,
        THREAD_LENGTH,
    )
    .expect("core should build")
}

fn thread() -> Solid {
    let bottom = -BODY_LENGTH - THREAD_LENGTH + THREAD_MARGIN;
    let turns = (THREAD_LENGTH - 2.0 * THREAD_MARGIN) / THREAD_PITCH;
    let axis = Axis3::new(Point3::new(0.0, 0.0, bottom), Vector3::z());
    let spine = edges::helix(
        axis,
        CORE_RADIUS,
        THREAD_PITCH,
        Rad64::new(0.0),
        Rad64::new(TAU * turns),
    )
    .expect("thread helix should build");

    let start = Helix::from_axis(axis, CORE_RADIUS, THREAD_PITCH).point_at(NativeParam::new(0.0));
    let outward = (start - axis.origin).normalize();
    let along = *axis.direction;

    let base = start - outward * THREAD_OVERLAP;
    let half = THREAD_BASE_WIDTH / 2.0;
    let profile = faces::polygon(&[
        base - along * half,
        start + outward * THREAD_DEPTH,
        base + along * half,
    ])
    .expect("thread profile should build");

    sweep_face(
        profile,
        &spine.edge(),
        SweepOptions {
            frame: SweepFrame::Axial(axis),
            samples_per_segment: (turns * SAMPLES_PER_TURN as f64).ceil() as usize,
            ..SweepOptions::default()
        },
    )
    .expect("thread should sweep")
}

fn main() {
    let counters = SolverCounters::snapshot();
    let started = Instant::now();
    let bolt = fuse(head(), body()).expect("head and body should fuse");
    let bolt = fuse(bolt, core()).expect("body and core should fuse");
    let bolt = fuse(bolt, thread()).expect("thread should fuse onto the core");
    let elapsed = started.elapsed();
    let counters = SolverCounters::snapshot().since(counters);

    validate_solid_orientation(bolt.model(), bolt.key()).expect("bolt should face outward");
    validate_solid_manifold(bolt.model(), bolt.key()).expect("bolt should be manifold");
    let volume = bolt
        .solid()
        .volume_properties()
        .expect("bolt volume should be measurable")
        .volume;
    println!("bolt built in {elapsed:.2?}, volume {volume:.6}");
    println!(
        "surface projections: {} global, {} hinted; branch fits {}",
        counters.global_surface_projections,
        counters.hinted_surface_projections,
        counters.branch_fits
    );
}
